File: sdk/js/test/library.test.ts
/**
 * Finding the addon: three positions, and the deliberate REVERSE precedence.
 *
 * `SHOJIKU_LIBRARY` beats explicit configuration, which beats the packaged
 * copy — the opposite of how the template root resolves, because where the
 * ENGINE lives is a deployment decision that has to be able to win over
 * application code.
 */

import { afterEach, describe, expect, it } from 'vitest';
import { Env } from '../src/env.js';
import {
  AbiMismatchError,
  Client,
  LibraryNotFoundError,
  resetConfiguration,
} from '../src/index.js';
import { ABI_VERSION, discover, PLATFORM_PACKAGES, packaged, requireAbi } from '../src/library.js';
import { Log } from '../src/log.js';
import { enginePath, makeClient } from './support/fixtures.js';

afterEach(resetConfiguration);

const silent = new Log();

describe('resolution order', () => {
  it('takes SHOJIKU_LIBRARY first, ahead of explicit configuration', () => {
    const env = new Env(true, { SHOJIKU_LIBRARY: '/from/env.node' });

    expect(discover('/from/config.node', env)).toEqual(['/from/env.node', 'environment']);
  });

  it('takes explicit configuration when the environment says nothing', () => {
    expect(discover('/from/config.node', new Env(true, {}))).toEqual([
      '/from/config.node',
      'configuration',
    ]);
  });

  it('falls back to the packaged addon when neither is given', () => {
    // No platform package is installed in this checkout, so the third position
    // is what it reports and the answer is "nothing found".
    expect(discover(null, new Env(true, {}))).toEqual([packaged(), 'packaged']);
  });

  it('ignores the environment entirely when lookups are disabled', () => {
    const env = new Env(false, { SHOJIKU_LIBRARY: '/from/env.node' });

    expect(discover('/from/config.node', env)).toEqual(['/from/config.node', 'configuration']);
  });
});

describe('the packaged position', () => {
  it('maps this machine to one platform package name', () => {
    expect(PLATFORM_PACKAGES.get('linux-x64')).toBe('@shojiku/linux-x64-gnu');
    expect(PLATFORM_PACKAGES.get('darwin-arm64')).toBe('@shojiku/darwin-arm64');
    expect(PLATFORM_PACKAGES.get('win32-x64')).toBe('@shojiku/win32-x64-msvc');
  });

  it('is nothing at all on a platform outside the matrix', () => {
    expect(packaged('sunos', 'sparc')).toBeNull();
  });

  it('is nothing when the package for this platform is not installed', () => {
    // The platform packages are published with the release; a source checkout
    // has none, which is exactly this case.
    expect(packaged('linux', 'x64')).toBeNull();
  });
});

describe('refusals', () => {
  it('names the install channels when no addon is found', () => {
    expect(() => new Client({ env: false })).toThrow(LibraryNotFoundError);
    try {
      // eslint-disable-next-line no-new
      new Client({ env: false });
    } catch (error) {
      const message = (error as Error).message;
      // A bare MODULE_NOT_FOUND names none of these, which is the whole reason
      // this error exists.
      expect(message).toContain('SHOJIKU_LIBRARY');
      expect(message).toContain('never downloads');
      expect(message).toContain('shojiku.node');
    }
  });

  it('names the path when the file is there but will not load', () => {
    expect(() => new Client({ library: '/etc/hostname', env: false })).toThrow(
      LibraryNotFoundError,
    );
  });

  it('refuses an addon whose ABI revision is not the one this package speaks', () => {
    // Split out from the constructor that feeds it: an addon linked against
    // this engine can only ever report the revision it was built with, so the
    // refusal is unreachable through a client and testable here.
    expect(() => requireAbi(ABI_VERSION + 1, '/some/addon.node', silent)).toThrow(AbiMismatchError);
  });

  it('admits the revision it does speak', () => {
    expect(() => requireAbi(ABI_VERSION, '/some/addon.node', silent)).not.toThrow();
  });
});

describe('a loaded library', () => {
  it('reports the path and which position won', () => {
    const client = makeClient();

    // Constructed with an explicit library and env off, so configuration is
    // the position that must win.
    expect(client).toBeInstanceOf(Client);
    expect(enginePath()).toContain('.node');
  });
});

File: sdk/java/src/test/java/jp/kengos/shojiku/LibraryTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.sun.jna.Native;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.jar.JarEntry;
import java.util.jar.JarOutputStream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

/**
 * Finding and opening the engine library, and the resolution order that is the deliberate reverse
 * of the template root's.
 */
class LibraryTest extends Fixtures {

  @Test
  void theEnvironmentBeatsExplicitConfigurationForTheLibrary() {
    // The reverse of the template root, on purpose: WHERE THE ENGINE LIVES is an
    // operator/deployment decision that has to be able to win over application
    // code — the same order the subprocess SDKs give SHOJIKU_BIN.
    Env env = new Env(true, Map.of("SHOJIKU_LIBRARY", LIBRARY));

    EngineLibrary library = new EngineLibrary("/nowhere/libshojiku_capi.so", env, new Log());

    assertEquals(LIBRARY, library.path());
    assertEquals("environment", library.source());
  }

  @Test
  void explicitConfigurationWinsWhenTheEnvironmentIsSilent() {
    EngineLibrary library = new EngineLibrary(LIBRARY, new Env(false), new Log());

    assertEquals(LIBRARY, library.path());
    assertEquals("configuration", library.source());
  }

  @Test
  void whichPositionWonIsReportedBecauseThatIsThe3amQuestion() {
    List<String> lines = new ArrayList<>();

    new EngineLibrary(LIBRARY, new Env(false), new Log(lines::add));

    assertTrue(
        lines.stream()
            .anyMatch(
                line -> line.contains("library_loaded") && line.contains("source=configuration")));
  }

  @Test
  void aLibraryThatIsNotThereNamesTheInstallChannels() {
    // The fix is always an installation step, and a bare loader error names none
    // of them.
    LibraryNotFoundException error =
        assertThrows(
            LibraryNotFoundException.class,
            () -> new EngineLibrary(null, new Env(false), new Log()));

    assertTrue(error.getMessage().contains("never downloads"));
    assertTrue(error.getMessage().contains("SHOJIKU_LIBRARY"));
  }

  @Test
  void aLibraryThatWillNotLoadNamesThemToo() throws IOException {
    Path notALibrary = Files.createTempFile("shojiku-not-a-lib", ".so");
    try {
      Files.writeString(notALibrary, "this is not an ELF object");

      LibraryNotFoundException error =
          assertThrows(
              LibraryNotFoundException.class,
              () -> new EngineLibrary(notALibrary.toString(), new Env(false), new Log()));

      assertTrue(error.getMessage().contains("could not be loaded"));
      assertTrue(error.getMessage().contains("never downloads"));
    } finally {
      Files.deleteIfExists(notALibrary);
    }
  }

  @ParameterizedTest
  // Windows is the reason there are six rather than three: cargo emits
  // `shojiku_capi.dll` with NO `lib` prefix while the Unix targets get one.
  // Looking only for the prefixed form would make the package unloadable on half
  // this SDK's market — one test per name, so a dropped candidate cannot hide
  // behind the others.
  @ValueSource(
      strings = {
        "libshojiku_capi.so",
        "shojiku_capi.so",
        "libshojiku_capi.dylib",
        "shojiku_capi.dylib",
        "libshojiku_capi.dll",
        "shojiku_capi.dll"
      })
  void everyPlatformsLibraryFilenameIsAmongTheCandidates(String name) {
    assertTrue(EngineLibrary.NAMES.contains(name));
  }

  @Test
  void theCandidateListHasNothingElseInIt() {
    // A negative sweep beside the positive one: a stray candidate would mean the
    // lookup probes for a filename no platform produces.
    assertEquals(6, EngineLibrary.NAMES.size());
  }

  @Test
  void thereIsNoPackagedBinaryInASourceCheckout() {
    // A test classpath is not a platform classifier jar, so there is nothing to
    // find — which is exactly the "no engine library was found" path.
    assertNull(EngineLibrary.packagedFrom(getClass().getClassLoader()));
  }

  @Test
  void aPackagedBinaryIsFoundOnTheClasspathWhenThereIsOne() throws IOException {
    // An UNPACKED classpath entry: the binary is already a file, so the lookup must
    // hand back that file rather than a copy of it.
    Path native0 = classpathRoot().resolve(EngineLibrary.PACKAGED_DIRNAME);
    Files.createDirectories(native0);
    Path planted = native0.resolve("libshojiku_capi.so");
    Files.copy(Path.of(LIBRARY), planted, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
    try {
      assertEquals(planted.toString(), EngineLibrary.packagedFrom(getClass().getClassLoader()));

      EngineLibrary library = new EngineLibrary(null, new Env(false), new Log());

      assertEquals(planted.toString(), library.path());
      assertEquals("packaged", library.source());
    } finally {
      Files.deleteIfExists(planted);
      Files.deleteIfExists(native0);
    }
  }

  @Test
  void theBinaryInsideARealPlatformJarIsExtractedAndLoadable() throws IOException {
    // The shape that actually SHIPS, and the one this suite used to miss: inside a
    // jar the binary is an archive entry, not a file, and `native/` names no
    // filesystem path at all. A lookup that resolved the directory passed every
    // test above and could not load one published artifact.
    Path jar = jarContaining("native/libshojiku_capi.so", Path.of(LIBRARY));
    try (URLClassLoader loader = new URLClassLoader(new URL[] {jar.toUri().toURL()}, null)) {
      String found = EngineLibrary.packagedFrom(loader);

      assertNotNull(found);
      // Extracted OUT of the archive: a path inside the jar would name nothing the
      // dynamic linker can open.
      assertFalse(found.contains(".jar!"));
      assertTrue(Files.isRegularFile(Path.of(found)));
      assertEquals(Files.size(Path.of(LIBRARY)), Files.size(Path.of(found)));

      // Whole bytes are not the claim that matters — loadability is. This is the
      // assertion the directory-shaped test could never make.
      assertEquals(
          EngineLibrary.ABI_VERSION,
          Native.load(found, ShojikuLibrary.class).shojiku_abi_version());
    } finally {
      Files.deleteIfExists(jar);
    }
  }

  @Test
  void aJarCarryingNoKnownBinaryNameResolvesToNothing() throws IOException {
    // A `native/` entry is where a platform jar PUTS the binary, not proof that one
    // is there. This is also what walks the whole candidate list to its end.
    Path jar = jarContaining("native/README.txt", null);
    try (URLClassLoader loader = new URLClassLoader(new URL[] {jar.toUri().toURL()}, null)) {
      assertNull(EngineLibrary.packagedFrom(loader));
    } finally {
      Files.deleteIfExists(jar);
    }
  }

  @Test
  void aPackagedDirectoryWithNoBinaryInItResolvesToNothing() throws IOException {
    // The unpacked twin of the case above: an empty `native/` must fall through to
    // the install hint, not to a path that opens nothing.
    Path native0 = classpathRoot().resolve(EngineLibrary.PACKAGED_DIRNAME);
    Files.createDirectories(native0);
    try {
      assertNull(EngineLibrary.packagedFrom(getClass().getClassLoader()));
      assertThrows(
          LibraryNotFoundException.class, () -> new EngineLibrary(null, new Env(false), new Log()));
    } finally {
      Files.deleteIfExists(native0);
    }
  }

  @Test
  void aLibrarySpeakingADifferentRevisionIsRefused() {
    // Loading anyway would mean calling symbols whose meaning has changed. The
    // rule is checked as a rule: a library that reports another revision cannot be
    // produced from the one this repository builds, and a refusal nobody can
    // exercise is a refusal nobody knows works.
    EngineLibrary.requireAbi(EngineLibrary.ABI_VERSION, "/lib.so");

    AbiMismatchException error =
        assertThrows(
            AbiMismatchException.class,
            () -> EngineLibrary.requireAbi(EngineLibrary.ABI_VERSION + 1, "/lib.so"));

    assertTrue(
        error.getMessage().contains("implements ABI revision " + (EngineLibrary.ABI_VERSION + 1)));
    assertTrue(error.getMessage().contains("speaks " + EngineLibrary.ABI_VERSION));
  }

  /**
   * A throwaway jar carrying one entry, which is the only way to test the shape that ships.
   *
   * @param entry the path inside the archive
   * @param content the file to store there, or null for an empty entry
   * @return the jar's path
   */
  private static Path jarContaining(String entry, Path content) throws IOException {
    Path jar = Files.createTempFile("shojiku-platform", ".jar");
    try (JarOutputStream out = new JarOutputStream(Files.newOutputStream(jar))) {
      out.putNextEntry(new JarEntry(entry));
      if (content != null) {
        Files.copy(content, out);
      }
      out.closeEntry();
    }
    return jar;
  }

  private static Path classpathRoot() {
    try {
      return Path.of(LibraryTest.class.getProtectionDomain().getCodeSource().getLocation().toURI());
    } catch (java.net.URISyntaxException error) {
      throw new IllegalStateException(error);
    }
  }

  @Test
  void theAbiRevisionIsCheckedBeforeAnythingElseIsCalled() {
    // Asked once: the only way a binding learns that a symbol it is about to call
    // means something different now.
    List<String> lines = new ArrayList<>();

    new EngineLibrary(LIBRARY, new Env(false), new Log(lines::add));

    assertEquals(1, EngineLibrary.ABI_VERSION);
    assertTrue(
        lines.stream()
            .anyMatch(line -> line.contains("abi_checked found=" + EngineLibrary.ABI_VERSION)));
  }
}
